src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php

```diff
@@ -673,6 +673,9 @@ public function create_item( $request ) {
 			);
 		}
 
+		// Don't save the post meta revision because they aren't stored yet.
+		remove_action( '_wp_put_post_revision', 'wp_save_revisioned_meta_fields', 10, 2 );
```

Comment:
I'm not sure I like this. The root of the problem is that the order of filters and actions in this REST API request is wrong. `wp_insert_post` and `wp_update_post` fire the `post_updated` action _after_ meta is updated. The REST API should do the same. Why does the REST API only fix the `wp_after_insert_post` to fire after meta is updated? It should do the same for `post_updated` and all those other filters.

Reply:
I don't know who else has recently worked on these endpoints. Maybe @TimothyBJacobs, @kadamwhite?

Reply:
I'm also not sure why the REST API decides not to use `meta_input`. Probably because it has limitation, but then those limitations should be addressed? Ideally everything goes through one code path. Since the REST API updates meta by itself, it had to move the `wp_after_insert_post` action, and will now have to do the same for all other actions.

So either we have to move those other actions, or use `meta_input` and address whatever the reasons were for not using that.

Reply:
The meta_input parameter is poorly designed for doing anything of complexity. It can't handle multi meta and support can't easily be added in a BC manner. It doesn't handle deleting meta. It doesn't provide any error handling. 

Perhaps the REST API could've been architected differently, but that ship has long since sailed. 

The REST API didn't move that hook, we introduced that hook specifically for the REST API. Core should be moving appropriate actions to that hook.

Reply:
It's not because something has been poorly designed that it can't be improved? Is there's something fundamentally unfixable about `meta_input`?

Core could move actions to that hook, but that ship has also sailed? We'd need to make sure plugins removing these hooks are still removing the moved hooks somehow.

Reply:
@ellatrix I also agree this solution isn't fantastic, it is a bit of a compromise approach. The one part I do like is hooking `wp_save_revisioned_meta_fields` on an action which lets any plugin easily disable the feature entirely.

I don't quite understand the part about `meta_input`.  Are you saying the REST endpoint could populate that before updating the post? What would be a more ideal solution in your mind?  

I am open to considering changing this - of course we would need to be careful not to break any cases where plugins already expect the current behavior. Also, this change could easily be done **after** this PR lands, so I don't feel it should be a blocker here. We would still hook the revisioning as is, we would just remove the special unhooking/re-hooking logic in the posts controller if the order was fixed.

Reply:
> I don't quite understand the part about `meta_input`. Are you saying the REST endpoint could populate that before updating the post? What would be a more ideal solution in your mind?

What I mean is that there's a `meta` property that can be passed to the post data on the `posts` endpoint, which is essentially the same as `meta_input` that can be passed to the post data for `wp_insert_post` and `wp_update_post`. So there's great potential for symmetry here. I don't get the argument of it being poorly designed for complex things. The input for both should be almost the same: a keys to values array. So you're telling me that whatever handling the REST API is doing for meta, cannot be done inside of `wp_update_post`? Why can't we add deletion and error handling? If we do that, the action order will be correct and all will be well?

Reply:
> So you're telling me that whatever handling the REST API is doing for meta, cannot be done inside of wp_update_post? Why can't we add deletion and error handling? If we do that, the action order will be correct and all will be well?

Hmmm good point! I will test out making the change you are suggesting, I'm not sure why that wouldn't be possible maybe I am missing something and @TimothyBJacobs can clarify. I do see that there are [other changes](https://github.com/WordPress/wordpress-develop/blob/2acd0aca7845feda1c4ca4dd708ca4e8a72c422b/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L899-L918C26) also completed after `wp_update_post`, but I don't see why we can't just include the meta in `wp_update_post` as `meta_input` and - as you say - _all will be well_.